### PR TITLE
CIP-0124? | Extend token metadata for translations

### PR DESCRIPTION
**Pull request by granada-pool**
This proposal defines an additional property to the NFT and FT metadata standard to support text localization.

---

([rendered proposal in branch](https://github.com/granada-pool/CIPs/blob/master/CIP-0124/README.md))

### CODE REVIEW

File: CIP-1856/README.md
Comment (rphair) [inline CIP-1856/README.md:3]:
```suggestion
Title: Extend metadata standard for translations
```
We need to keep these titles short and clear.  If there is a "token" that is not either an NFT or FT, please correct me.  Just suggesting this as an average between the PR title and the CIP title.  Long history of EE/CS standard names is to avoid punctuation wherever possible.

File: CIP-1856/README.md
Comment (rphair) [inline CIP-1856/README.md:134]: Thanks for adding this section with the relevant points, but could you add something about who will be implementing this... even if you don't have any specific parties yet who are working on acceptance & implementation?  If you've talked with anyone about the need for this standard, or made some progress yourselves, it would be helpful to reflect that here.

**Review by rphair — COMMENTED** [review]
Please rename your project folder from `CIP-1856` to something else beginning with `CIP-` but not a number (derived from your proposal name, or `????`).

This is relatively important for `1856` because we had a potential CIP-1856 for a while: https://github.com/cardano-foundation/CIPs/pull/104

File: CIP-1856/README.md
Comment (granada-pool) [reply CIP-1856/README.md:134]: @rphair thank you very much for the suggestions and refinements! I'm in contact with Patrick Tobler from NMKR to see if we can implement this together (not confirmed yet). I'm also creating a confidential NFT launch pad on the Oasis Network called geniish.io (the idea is to bridge it to Cardano as soon as it's technically feasible with the sidechain kits), so I came up with the idea in collaboration with my business partner Aaron Schmid and my co-SPO Carolina Isler (LaPetiteADA). I also have 6 years of experience in the translation and copywrite branch, so I believe that web3 will follow the same steps from most web2 giants as localization is the key to a globalized market. I'll add this section as soon as I have confirmed the partnerships.

File: CIP-1856/README.md
Comment (SebastienGllmt) [inline CIP-1856/README.md:48]: Probably it's best to update the cddl definition of CIP25 instead of a JSON specification like this

Comment (SebastienGllmt) [reply CIP-1856/README.md:48]: https://github.com/cardano-foundation/CIPs/blob/master/CIP-0025/cddl/version_2.cddl

File: CIP-1856/README.md
Comment (granada-pool) [reply CIP-1856/README.md:136]: @rphair So NMKR has provided valuable peer feedback and they would like to support this CIP as well. I have included this updates on the proposal.

**Comment by rphair** [comment]
Thanks @granada-pool for the latest updates & including the peer review from NMKR.  Can you please [rename the project folder](https://github.com/cardano-foundation/CIPs/pull/488#pullrequestreview-1356889657) from `CIP-1856` to something without a number it it (e.g. `CIP-????`, `CIP-XXXX`, `CIP-token-metadata-translations`)?

File: CIP-1856/README.md
Comment (granada-pool) [reply CIP-1856/README.md]: Good idea, I'll fork that CIP and will add the new attributes once this proposal is approved.

File: CIP-token-metadata-translations/README.md
Comment (rphair) [inline CIP-token-metadata-translations/README.md:147]:
```suggestion
NMKR has supported this CIP with peer feedback, and in the future with a pilot implementation of this localization method.
```
@granada-pool I don't think we can use the term "standard" here because of the multiple ways this problem might be addressed.  The above is one idea which would also be specific about what would be implemented.

**Review by rphair — COMMENTED** [review]
@Padierfind if you could please assist here:
- This proposal seems good enough for me to approve (mainly based on web standards experience; I'm not a token metadata expert) if we could have a peer review, or even just an acknowledgement, of this PR coming from the NMKR direction.
- If there are any revisions to how your peer review is presented in the CIP please suggest alternative language and/or discuss it in this PR.

File: CIP-token-metadata-translations/README.md
Comment (granada-pool) [reply CIP-token-metadata-translations/README.md:147]: I agree! There are indeed several methods to solve this. Although, I really liked the idea of being a bit "aggressive" or "arrogant" and call it a **standard**, just like Ethereum took the liberty of creating the _ERC721 Metadata Standard_ for their NFTs, which has been consolidated by [OpenSea ](https://docs.opensea.io/docs/metadata-standards), and it's even used in Cardano's NFTs. Based on my experience in web2 text localization, this approach is quite solid, and since Cardano would be the first chain ever addressing this language barrier in the space, why not call it a _Cardano standard_? Specially, if NMKR will eventually become larger than OpenSea. 😉

Comment (rphair) [reply CIP-token-metadata-translations/README.md:147]: Then indeed we could wait until we see other community members publicly calling it a "standard" as was the case with ERC721.  As it is now this is just the most promising _candidate_ for a localisation standard. 😉

**Comment by granada-pool** [comment]
Hey guys, any updates on this proposal?

File: CIP-token-metadata-translations/README.md
Comment (rphair) [inline CIP-token-metadata-translations/README.md:145]:
```suggestion
Propose this method in documentation and references for web3 developers.

```
1. Here and elsewhere I would suggest avoiding the term "standard" since only adoption & longevity can characterise it as such.  Until that time comes, this is only a "method" or a "proposal".
2. All items in the Path to Active would work well as tickboxes; see the GitHub formatting and/or more recent CIPs to see how we've been doing these, which would suit your CIP's progression from `Proposed` to `Active` as they are ticked,

File: CIP-token-metadata-translations/README.md
Comment (rphair) [inline CIP-token-metadata-translations/README.md:147]: @Padierfind (if you are still the appropriate contact for NMKR) please validate this if you can, along with any other contributions you can make to the review of this proposal (echoing my earlier request from https://github.com/cardano-foundation/CIPs/pull/488#pullrequestreview-1398114410).

**Review by rphair — COMMENTED** [review]
@granada-pool in my last request https://github.com/cardano-foundation/CIPs/pull/488#pullrequestreview-1398114410 I'd been hoping for some community endorsement of your proposed method: hence the wait for something else to happen either from the community or from you.

Editorially your proposal still seems good as a suggested method... especially since no other applicable method has been suggested here in the last year or more.  Let me suggest some format corrections and then we'll see if the other active editors @Ryun1 @Crypto2099 also feel the same about assigning a CIP number to this one & getting it merged around the next CIP meeting or two.

In any case I've put it up for review at the next CIP meeting so we can all have a second look at it together & move forward: https://hackmd.io/@cip-editors/90

**Review by rphair — COMMENTED** [review]
@granada-pool I've taken the liberty of https://github.com/cardano-foundation/CIPs/pull/488/commits/d98440167087e8cd9a1778064a2c79442a4ef8b1 rather than rounds of proposed changes: cleaning up some redundant markup in the Copyright section and reorganising & marking up your _Path to Active_ with tickboxes and rephrasing as goals which editors and the community can verify.

Please double check my changes for this commit and especially confirm which boxes should currently be ticked.  Since there is no feedback from @Padierfind or NMKR associates you may have to provide some evidence of their involvement (other than the "feedback" you mention) if you want to cite their implementation of your proposal.

**Review by VitoMelchionna — APPROVED** [review]

**Comment by VitoMelchionna** [comment]
> @granada-pool I've taken the liberty of [d984401](https://github.com/cardano-foundation/CIPs/commit/d98440167087e8cd9a1778064a2c79442a4ef8b1) rather than rounds of proposed changes: cleaning up some redundant markup in the Copyright section and reorganising & marking up your _Path to Active_ with tickboxes and rephrasing as goals which editors and the community can verify.
> 
> Please double check my changes for this commit and especially confirm which boxes should currently be ticked. Since there is no feedback from @Padierfind or NMKR associates you may have to provide some evidence of their involvement (other than the "feedback" you mention) if you want to cite their implementation of your proposal.

Sounds good! I have contacted @Padierfind, but I haven't gotten an answer yet.

Meanwhile, the best evidence I can provide is our chat on X (in German):
WhatsApp Image 2024-06-07 at 12 59 16
WhatsApp Image 2024-06-07 at 12 59 16 (1)

Patrick suggested keeping the translations off-chain (similarly to ETH), as they can easily surpass the memory limits of Cardano, which was shortly added to the proposal after the exchange of ideas.

**Comment by granada-pool** [comment]
> @granada-pool I've taken the liberty of [d984401](https://github.com/cardano-foundation/CIPs/commit/d98440167087e8cd9a1778064a2c79442a4ef8b1) rather than rounds of proposed changes: cleaning up some redundant markup in the Copyright section and reorganising & marking up your _Path to Active_ with tickboxes and rephrasing as goals which editors and the community can verify.
> 
> Please double check my changes for this commit and especially confirm which boxes should currently be ticked. Since there is no feedback from @Padierfind or NMKR associates you may have to provide some evidence of their involvement (other than the "feedback" you mention) if you want to cite their implementation of your proposal.

Sounds good! I have contacted @Padierfind, but I haven't gotten an answer yet.

Meanwhile, the best evidence I can provide is our chat on X (in German):
WhatsApp Image 2024-06-07 at 12 59 16
WhatsApp Image 2024-06-07 at 12 59 16 (1)

Patrick suggested keeping the translations off-chain (similarly to ETH), as they can easily surpass the memory limits of Cardano, which was shortly added to the proposal after the exchange of ideas.

File: CIP-token-metadata-translations/README.md
Comment (Ryun1) [inline CIP-token-metadata-translations/README.md:30]: As this contains on-chain data, I believe it must also have a CDDL specification (see [here](https://github.com/cardano-foundation/CIPs/blob/master/CIP-0001/README.md?plain=1#L83C541-L83C620)).

File: CIP-token-metadata-translations/README.md
Comment (Ryun1) [inline CIP-token-metadata-translations/README.md:118]:
```suggestion
     
## Rationale: how does this CIP achieve its goals?

```
Please add a rationale section!

File: CIP-token-metadata-translations/README.md
Comment (Ryun1) [inline CIP-token-metadata-translations/README.md:3]:
```suggestion
Title: CIP-25 - Extend token metadata for translations
```

might make sense to do something like this, as this is a strictly CIP25 extension

File: CIP-token-metadata-translations/README.md
Comment (Ryun1) [inline CIP-token-metadata-translations/README.md:24]: Probably want to be specific here, for which version of CIP-25 this applicable to https://github.com/cardano-foundation/CIPs/tree/master/CIP-0025#cddl

File: CIP-token-metadata-translations/README.md
Comment (rphair) [inline CIP-token-metadata-translations/README.md:2]:
```suggestion
CIP: 124
```

**Review by rphair — COMMENTED** [review]
@granada-pool this was generally seen at tonight's CIP meeting to be ready to progress to the final `Last Check` stage before merge as soon as @Ryun1's feedback above is addressed.  (The CDDL spec is a rather big thing that it needs first.)

Please also change the containing directory name to `CIP-0124` 🎉

**Comment by rphair** [comment]
@granada-pool are you still advocating this?  Based on no response to the 2 reviews above it appears abandoned & so we're tagging at such (it's not really `Waiting for Author` since the proposal specification was well received in review except for some document particulars, as recorded above).

It would seem a shame to close this PR after you have basically done all the work, so please respond to the feedback above if & when you can.  (Proposals marked as apparently "Abandoned" will be closed some time afterward if no further progress is made.)

File: CIP-token-metadata-translations/README.md
Comment (VitoMelchionna) [reply CIP-token-metadata-translations/README.md:24]: Both versions. This CIP would improve the metadata section.

**Review by VitoMelchionna — APPROVED** [review]

**Comment by granada-pool** [comment]
> @granada-pool this was generally seen at tonight's CIP meeting to be ready to progress to the final `Last Check` stage before merge as soon as @Ryun1's feedback above is addressed. (The CDDL spec is a rather big thing that it needs first.)
> 
> Please also change the containing directory name to `CIP-0124` 🎉

@rphair Great to know! I just made some changes based on your comments and suggestions

**Comment by granada-pool** [comment]
> @granada-pool are you still advocating this? Based on no response to the 2 reviews above it appears abandoned & so we're tagging at such (it's not really `Waiting for Author` since the proposal specification was well received in review except for some document particulars, as recorded above).
> 
> It would seem a shame to close this PR after you have basically done all the work, so please respond to the feedback above if & when you can. (Proposals marked as apparently "Abandoned" will be closed some time afterward if no further progress is made.)

@rphair Absolutely! I apologize for the delay, I had quite a lot on my plate with the development of VAULT3 and the RareEvo event. Hope the changes made today satisfy all reviews. Please tag me if there are more improvements to make.

**Review by rphair — COMMENTED** [review]
thanks @granada-pool - just one last pass for me before we try to push this through:

File: CIP-0124/README.md
Comment (rphair) [inline CIP-0124/README.md:3]:
```suggestion
Title: Extend token metadata for translations
```
Would be better to keep the long time PR title for now since that's the way reviewers were used to seeing this & there's never been an ambiguity with this being applicable to more modern token types.  And someday it might have a CIP modification that will make it so... therefore to avoid fragmentation it would help to have a more general title.

In any case the other CIP name cannot be in the initial position because it at first glance it appears to be the "wrong" CIP 😅

File: CIP-0124/README.md
Comment (rphair) [inline CIP-0124/README.md:90]:
```suggestion
## References
```
Please move this section to the end by the convention that evolved during this PR's lifetime: at H2 level (`##`) and before the Copyright.  (see CIP-0001 > optional sections)

**Review by rphair — COMMENTED** [review]
(accidentally hit "Submit" button above)

File: CIP-token-metadata-translations/README.md
Comment (granada-pool) [reply CIP-token-metadata-translations/README.md]: Done!

File: CIP-0124/README.md
Comment (granada-pool) [reply CIP-0124/README.md:90]: I followed the structure of [CIP-25](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0025), but I can move this section as requested.

Comment (granada-pool) [reply CIP-0124/README.md]: Done! ;-)

**Review by rphair — APPROVED** [review]
@granada-pool I think we are all in order now although I will leave it to @Ryun1 @Crypto2099 to confirm the CDDL spec + recent updates.  Marking as `Last Check` since I think this is now good to finally merge as soon as we have their confirmation.

File: CIP-0124/README.md
Comment (rphair) [reply CIP-0124/README.md]: @granada-pool I believe it should also be changed in CIP-0025.  At the time I re-reviewed CIP25 at the beginning of the year, it seemed its Reference applied directly to the spec above it... now I can see they're more applicable to the document as a whole.  Thanks for pointing that out & we'll attend to the CIP25 structure at the next opportunity.

File: CIP-0124/README.md
Comment (Ryun1) [inline CIP-0124/README.md:90]:
```suggestion
### CDDL
```

**Review by Ryun1 — APPROVED** [review]

**Pull request merged**
